src/nspawn/nspawn-seccomp.c

```diff
@@ -29,111 +29,154 @@
-                { CAP_SYS_PACCT,  SCMP_SYS(acct)                },
-                { CAP_SYS_PTRACE, SCMP_SYS(process_vm_readv)    },
-                { CAP_SYS_PTRACE, SCMP_SYS(process_vm_writev)   },
-                { CAP_SYS_PTRACE, SCMP_SYS(ptrace)              },
-                { CAP_SYS_RAWIO,  SCMP_SYS(ioperm)              },
-                { CAP_SYS_RAWIO,  SCMP_SYS(iopl)                },
-                { CAP_SYS_RAWIO,  SCMP_SYS(pciconfig_iobase)    },
-                { CAP_SYS_RAWIO,  SCMP_SYS(pciconfig_read)      },
-                { CAP_SYS_RAWIO,  SCMP_SYS(pciconfig_write)     },
-#ifdef __NR_s390_pci_mmio_read
-                { CAP_SYS_RAWIO,  SCMP_SYS(s390_pci_mmio_read)  },
-#endif
-#ifdef __NR_s390_pci_mmio_write
-                { CAP_SYS_RAWIO,  SCMP_SYS(s390_pci_mmio_write) },
-#endif
-                { CAP_SYS_TIME,   SCMP_SYS(adjtimex)            },
-                { CAP_SYS_TIME,   SCMP_SYS(clock_adjtime)       },
-                { CAP_SYS_TIME,   SCMP_SYS(clock_settime)       },
-                { CAP_SYS_TIME,   SCMP_SYS(settimeofday)        },
-                { CAP_SYS_TIME,   SCMP_SYS(stime)               },
+                const SyscallFilterSet set;
+        } cap_filter_sets[] = {
+                { CAP_SYS_MODULE,       { .value = "@module\0"  }},
+                { CAP_SYS_RAWIO,        { .value = "@raw-io\0"  }},
+                { CAP_SYS_TIME,         { .value = "@clock\0"   }},
+                { CAP_SYSLOG,           { .value = "syslog\0"   }},
+                { CAP_SYS_PACCT,        { .value = "acct\0"     }},
+                { CAP_SYS_PTRACE,       { .value =
+                                        "process_vm_readv\0"
+                                        "process_vm_writev\0"
+                                        "ptrace\0"              }},
         };
+        int r;
         unsigned i;
-        int r, c = 0;
 
-        for (i = 0; i < ELEMENTSOF(blacklist); i++) {
-                if (blacklist[i].capability != 0 && (cap_list_retain & (1ULL << blacklist[i].capability)))
+        for (i = 0; i < ELEMENTSOF(cap_filter_sets); i++) {
+                /* const char *t; */
+
+                /* skip if we need to retain the capability */
+                if (cap_list_retain & (1ULL << cap_filter_sets[i].capability))
                         continue;
 
-                r = seccomp_rule_add_exact(ctx, SCMP_ACT_ERRNO(EPERM), blacklist[i].syscall_num, 0);
-                if (r < 0) {
-                        /* If the system call is not known on this architecture, then that's fine, let's ignore it */
-                        _cleanup_free_ char *n = NULL;
+                r = seccomp_filter_set_add(blacklist, !whitelist, &(cap_filter_sets[i].set));
+                if (r < 0)
+                        log_debug_errno(r, "Failed to add capability to blacklist, ignoring: %m");
+                /* NULSTR_FOREACH(t, cap_filter_sets[i].set.value) { */
```

Comment:
I wasn't sure if the custom filter should also whitelist syscalls from the capability filter sets. I'm not sure what the priorities are there.

Reply:
One angle would be that we should allow to whitelist those syscalls so apps can get a proper EACCESS or whatever instead of the seccomp blocking behavior, whatever that is.

Other angle would be that if the user didn't want to keep certain capability, restricting the syscalls adds a bit more security in the case there's a bug in the capabilities implementation.

I think that if the user doesn't enable a capability but enables the syscalls it's their fault and they should be allowed to shoot themselves on the foot :)

TL;DR whitelisting is fine IMO